### ISSUE TITLE
[feature request] - Replacement for Algolia Places Required

### ISSUE THREAD

**Issue by chrismclarke** [ONEARMY/community-platform#1177]
**Is your feature request related to a problem? Please describe.**
We currently use algolia places to autocomplete location search on map and user profile pages, however Algolia have announced they are _sunsetting_ the project and will no longer be supporting after May 31st 2022 (and have already ceased new signups)
https://www.algolia.com/blog/product/sunseting-our-places-feature/

**Describe the solution you'd like**
Ideally we want to find either a similar commercial solution that has a decent free tier, or an open-source solution that could be self-hosted (or paid as a service temporarily). Ideally it will continue to use Open Streets Map data for consistency with the data that already exists (and avoiding google map lockin). A couple platforms Algolia mentions are:
- [Mapbox](https://www.mapbox.com/)
- [Google Places](https://cloud.google.com/maps-platform/places)
- [Geocode Earth](https://geocode.earth/)

A couple others that came up with a quick search on github include:
- [Photon](https://photon.komoot.io/)
- [OSM Names](https://osmnames.org/)
- [Pelias](https://pelias.io/)

Not sure if anybody has used any of these before or has any other suggestions?

**Additional context**

_Example of Algolia Places used on the map page_
![image](https://user-images.githubusercontent.com/10515065/123313694-9bffe880-d4de-11eb-8509-ecf6cab9b749.png)

**Comment by davehakkens** [comment]
came across this map. Looks cool. 
Just dumping it here for alternative inspiration
https://www.beehome.design/community/

**Comment by chrismclarke** [comment]
Thanks @thisislawatts for reviving this on the call today.

I did a quick bit more investigating and of the 6 options mentioned GeoCode earth is built on top of Pelias, so could be a self-hosted option (albeit they recommend some pretty substantial hardware requirements for hosting which would need to be reviewed).

I think to move things forward the best thing to do would be @davehakkens , if either you or anyone in the community could look at the websites of the different options and test the various demos. It would be good to know if any perform better/worse when searching for workspace locations. Find direct links below:

Mapbox - https://docs.mapbox.com/playground/geocoding/
Google Places - https://developers.google.com/maps/documentation/javascript/examples/places-autocomplete
Geocode Earth / Pelias - https://geocode.earth/docs/forward/autocomplete/

Photon - https://photon.komoot.io/
OSM Names - https://osmnames.org/

**Comment by orangejulius** [comment]
Hi folks,

Co-maintainer of Pelias and co-founder of Geocode Earth here. I check Github for mentions of our work pretty carefully and came across this discussion :)

We often provide free plans for non-profits and other organizations. If self-hosting Pelias is something you're looking at we can also provide a few tips for doing that as cost effectively as possible.

Feel free to shoot us an email if we can help.

**Comment by davehakkens** [comment]
haha hey @orangejulius. Thanks for popping in here.
Do you think Pelias would be a good fit for our map?
https://community.preciousplastic.com/map

**Comment by thisislawatts** [comment]
Reading through [the documentation around full planet] builds of Pelias, I am not sure a self-hosted instance is a good fit given our project constraints. Managing an ElasticSearch install with more than 400Gb is going to require more resources (time/money) than we currently have available. I think a plan with the Geocode Earth, would be a better fit.

**Comment by orangejulius** [comment]
@davehakkens yes definitely. The autocomplete functionality from Pelias would fit great there.

Setting up Pelias for a full planet with _all_ the data can be rather costly, but if you want to limit to something like cities only it's pretty manageable. I want to say around 10GB worth of data. Depending on your needs that would certainly be an option.

**Comment by kfern** [comment]
Hi!

Could we use open data sets like those in https://www.higithub.com/planetopendata/repo/awesome-world or similar?

The place with highest number of sites on the map is the north of France, with 60.

Do we need to search by location or would be enough by county? 

If we do the search by country it would not be necessary to use an external service.

This search only changes the center on the map.

**Comment by davehakkens** [comment]
@b-n any suggestions on this? 👆
As a PP map veteran

**Comment by b-n** [comment]
@davehakkens Now you're making me feel old! (and too right, those maps had me beat in 'nam!)

TL;DR: Ditch street level search, look for city/country level search because that's what people probably use it for anyway.

I would go way back to the question which to me is: "How do people use search on the map? And what are they expecting the map to do?

I imagine it's a mix of people that search by:
- City
- Country
- Postcode
- ~Their own address?~ <-- highly doubt it, but I'm sure they'd fall back to city if not available

I imagine people use the search as a fuzzy match for an area, and zoom in/out to discover places around the area. e.g. Street level search is kind of redundant. And let's throw post codes into that same bucket. It's "handy", but just as handy as "click to share your location" sort of buttons which (if not already there, I think it is), fulfill that problem. You will be one of the few users that actually searches outside of their location, because you visit other countries etc.

Okay, so City/Country search. And remembering this is just to zoom the user to the location on the map more than anything else (e.g. Auckland throws you to a GPS point in NZ).

There are apparently ~10,000 cities in the world. I doubt there are new ones daily. The issue with cities, is that we don't all speak the same language, so you can multiple that by a bunch more for different scripts if you're concerned about non-latin texts. That's a not-insignificant amount of data to keep up to date, and thus why people/companies like @orangejulius exist - because cities would be considered as easy as melting butter in the Sahara in the middle of a cloud free sunny day - in comparison to street level stuff. Elastic search, and self hosting that dataset is where that 10gb comes in, I assume that's the size of the cities/geolocation index. To be fair, people don't pay Geocode Earth for the tool, they're paying for access to the data normally (keeping up to date can only ever be so automated, and anything that isn't automated = $$$ af)

Okay, so then: Do you __really__ want cities in the drop down is the big question.

If yes, Geocode Earth sounds fun - but I have no idea of the support costs, maintenance etc since the cost of anything is not the install - but what comes later with keeping up to date.
Other options, maybe OSMNames, but you will get 0 support from them directly, and it'll be hodge podging a bunch of stuff together.

Okay okay, so if that all feels like much, I have maybe the easiest solution of them all. A country picklist, and you just manage a list of the countries to their respective GPS points (and maybe allow people to update said JSON file for their own scripts/dialects). Is it pretty, and awesome? Nope. Is it hella functional and saves a bunch of development effort? Yes. (Is it somewhat bad in terms of data usage because people have to download the index when the page loads? Okay yes, but it's not "that" much data)

**Comment by davehakkens** [comment]
thanks for the reply @b-n 
I think a country dropdown on the front-end could actually be enough, specially knowing it could reduce complexcity.
Most users don't have specific searches but browse areas.

But how would this work for a user to set a pin? You zoom in to your location? 
Could see this being challenging in a small profile-admin-screen....

![afbeelding](https://user-images.githubusercontent.com/13672737/146346843-04329a49-e7e7-488b-8f79-39f787c5f39e.png)

**Comment by kfern** [comment]
HI @davehakkens 

This use case appears to meet the "Acceptable Use Policy for the server running at nominatim.openstreetmap.org" requirements for free API access.
https://operations.osmfoundation.org/policies/nominatim/

API Playground
https://nominatim.openstreetmap.org/ui/search.html

**Comment by b-n** [comment]
@davehakkens Yeah, that would be slightly more ideal to have address. But even then, most users do it once, and they generally know where they are situated, so it should be "fairly fast" for them to do it and forget that the "bad" UX ever existed. Could just scale the map to be bigger there too :smile: . Could also throw a "zoom to me" button that asks if the browser can have the users location for at least getting the map in generally the right area.

Or alternatively, the OSM public API as above from @kfern for the input form. I'd avoid using it on the main map page - purely because that's extra load for the OSM free service people, and as you said - people just browse anyway and a country is nice there.

**Comment by thisislawatts** [comment]
We still have some until Algolia Places is sunset entirely, I think it would be worth introducing some analytics on the LocationSearch component so we can understand how folks are using it.

It may be that we end up splitting the implementation out to a few different components which are more focused on the specific user needs.

Based on the thread so far, I see there being two distinct use cases:

**1. Address to Lat/Long**
We want fuzzy human addresses to be turned into lat/long co-ordinates. As @b-n highlights this is likely used much less than the search on our public facing Maps page so we **may** be able to piggyback on OSM's free API here. 

**2. City/Country to Pin**
Search all **existing** pins using countries/cities as a filtering mechanism. So the autosuggest returns any pins which match against the queried City or Country names. I think this would allow us to entirely sidestep a degree of complexity here. Clicking one of the suggested Pins would update the maps location then allowing folks to visually browse the local area. 

See attached screengrab for an example of how a search for a location not covered by our existing map pins would appear. 

![Screenshot 2021-12-28 at 12 48 23](https://user-images.githubusercontent.com/472589/147563371-bc8c427f-2dea-4923-8aa9-4953564af50b.jpg)

---

See attached screengrab for an example of how a location search which matches against Pins in the database. 

![Screenshot 2021-12-28 at 12 56 13](https://user-images.githubusercontent.com/472589/147563966-75cfb712-f933-4fc5-91a2-da2f4478296e.jpg)

**Comment by thisislawatts** [comment]
Hallo, following up on the event we introduced to try and understand the usage of our search tool. Based on the 2032 events we have seen in the last 16 days I think we currently sit within the definition of "reasonable usage" for the OSM API. 

We average **1 request for every 680 seconds**, although this rough calculation ignores:
* Clustering of events, for example our highest single day had 213 events. If all of these events had occurred within a single 6 minute window would exceed the terms of a reasonable usage agreement. 
* Inaccuracies in the data 😬 People using adblockers or do not track settings on their browser will not be included in the data.

As a result of the above, I would recommend:
1. Moving forward with OSM as the replacement for Algolia
2. Implement a robust failure mode if the OSM endpoint is unavailable due to rate limiting.
3. Send a recurring/one-off donation or create a membership account to support the OSM API project. https://wiki.osmfoundation.org/wiki/Donate

**Comment by davehakkens** [comment]
Where do you find these stats, or how do we know we go beyond "reasonable usage" in the future?
But generally i like the idea, seems like a smooth option. 
Also up to support with a donation, their maps have been very useful.

**Comment by thisislawatts** [comment]
All of these events are currently being captured and logged to Google Analytics, so we can check the data there although there is a slight delay on items being visible in the reports.

As part of: _2. Implement a robust failure mode if the OSM endpoint is unavailable due to rate limiting._ I believe that we should set up some specific alerting around rate limits being reached. Either using Sentry or another tool.
